src/overlays/actors/ovl_En_Bubble/z_en_bubble.c

```diff
@@ -23,45 +26,393 @@ const ActorInit En_Bubble_InitVars = {
+    if (DECR(this->explosionCountdown) != 0) {
+        return -1;
+    }
+    return func_809CBCEC(this);
+}
+
+// only called in an unused actionFunc
+u32 func_809CC020(EnBubble* this) {
+    this->expansionWidth += 0.083333336f;
+    this->expansionHeight += 0.083333336f;
+
+    if (DECR(this->explosionCountdown) != 0) {
+        return 0;
+    }
+    return 1;
+}
+
+void EnBubble_Vec3fNormalizedRelfect(Vec3f* vec1, Vec3f* vec2, Vec3f* ret) {
+    f32 norm;
+
+    Math3D_Vec3fReflect(vec1, vec2, ret);
+    norm = sqrtf((ret->x * ret->x) + (ret->y * ret->y) + (ret->z * ret->z));
+    if (norm != 0.0f) {
+        ret->x /= norm;
+        ret->y /= norm;
+        ret->z /= norm;
+    } else {
+        ret->x = ret->y = ret->z = 0.0f;
+    }
+}
+
+void EnBubble_Vec3fNormalize(Vec3f* vec) {
+    f32 norm;
+
+    norm = sqrt((vec->x * vec->x) + (vec->y * vec->y) + (vec->z * vec->z));
+    if (norm != 0.0f) {
+        vec->x /= norm;
+        vec->y /= norm;
+        vec->z /= norm;
+    } else {
+        vec->x = vec->y = vec->z = 0.0f;
+    }
+}
+
+void EnBubble_Fly(EnBubble* this, GlobalContext* globalCtx) {
+    CollisionPoly* sp94;
+    Actor* bumpActor;
+    Vec3f sp84;
+    Vec3f sp78;
+    Vec3f sp6C;
+    Vec3f sp60;
+    Vec3f sp54;
+    f32 bounceSpeed;
+    u32 sp4C;
+    u8 bounceCount;
+
+    if (this->colliderSphere.list[1].body.bumperFlags & 0x2) {
+        bumpActor = this->colliderSphere.base.ac;
+        this->normalizedBumpVelocity = bumpActor->velocity;
+        EnBubble_Vec3fNormalize(&this->normalizedBumpVelocity);
+        this->velocityFromBump.x += (this->normalizedBumpVelocity.x * 3.0f);
+        this->velocityFromBump.y += (this->normalizedBumpVelocity.y * 3.0f);
+        this->velocityFromBump.z += (this->normalizedBumpVelocity.z * 3.0f);
+    }
+    this->sinkSpeed -= 0.1f;
+    if (this->sinkSpeed < this->actor.minVelocityY) {
+        this->sinkSpeed = this->actor.minVelocityY;
+    }
+    sp54.x = this->velocityFromBounce.x + this->velocityFromBump.x;
+    sp54.y = this->velocityFromBounce.y + this->velocityFromBump.y + this->sinkSpeed;
+    sp54.z = this->velocityFromBounce.z + this->velocityFromBump.z;
+    EnBubble_Vec3fNormalize(&sp54);
+
+    sp78.x = this->actor.posRot.pos.x;
+    sp78.y = this->actor.posRot.pos.y + this->actor.shape.unk_08;
+    sp78.z = this->actor.posRot.pos.z;
+    sp6C = sp78;
+
+    sp6C.x += (sp54.x * 24.0f);
+    sp6C.y += (sp54.y * 24.0f);
+    sp6C.z += (sp54.z * 24.0f);
+    if (func_8003DE84(&globalCtx->colCtx, &sp78, &sp6C, &sp84, &sp94, 1, 1, 1, 0, &sp4C) != 0) {
+        sp60.x = sp94->norm.x * 0.00003051851f;
+        sp60.y = sp94->norm.y * 0.00003051851f;
+        sp60.z = sp94->norm.z * 0.00003051851f;
+        EnBubble_Vec3fNormalizedRelfect(&sp54, &sp60, &sp54);
+        this->bounceDirection = sp54;
+        bounceCount = this->bounceCount;
+        this->bounceCount = ++bounceCount;
+        if (bounceCount > (s16)(Rand_ZeroOne() * 10.0f)) {
```

Comment:
Can this variable be avoided by any chance? Maybe doing this or inlining `++this->bounceCount` in the condition?
```suggestion
        this->bounceCount++;
        if (this->bounceCount > (s16)(Rand_ZeroOne() * 10.0f)) {
```

Reply:
I messed around a lot with this, its the best I could get that matched. it is weird though